firecracker/kernel/build.sh

```diff
@@ -45,3 +47,15 @@ tar \
     --create \
     --gzip \
     "${KERNEL_BIN_FILE}"
+
+########################################
+# Write a manifest file containing the version.
+# The version should dedupe based on input file checksums.
+# (This may be replaced with Pants-based file diffs in the future.)
+########################################
+source .buildkite/scripts/lib/manifest.sh
+INPUTS_SHA="$(sha256_of_dir firecracker/kernel)"
+readonly INPUTS_SHA_SHORT="${INPUTS_SHA:0:16}" # cloudsmith version field must be under 128 chars
+readonly MANIFEST_PATH="${DISTRIBUTION}.manifest"
+readonly VERSION="firecracker-${FIRECRACKER_RELEASE}-kernel-${KERNEL_VERSION}-input-sha256-${INPUTS_SHA_SHORT}"
```

Comment:
🤔 This version seems rather unwieldy, and includes extra information that is really just an "implementation detail" to get around the fact that we don't have a deterministic build.

The fact that we're also making a JSON object to just store this string seems overkill, and a bit limiting.

My suggestion would be to keep the version as it is currently (`firecracker-${FIRECRACKER_RELEASE}-kernel-${KERNEL_VERSION}`), and also keep the full-length SHA256 "input checksum". The JSON would look like this:

```json
{
    "version": "firecracker-XXX-kernel-YYY",
    "input-sha256": "abcdef01234656789..."
}
```
(Ideally, if this "manifest" is going to stay around long-term and also be uploaded to Cloudsmith as an artifact in and of itself, I'd break out the Firecracker version and the kernel version into their own dedicated fields, but that's a conversation for another time. It would also need a bit more though to harmonize with the usage for the rootfs image.)

Then, when you upload, you can have access to all the raw information and can do with it what you please. In particular, I think expressing this whole input checksum thing as a _tag_ would be a bit better and more flexible of a structure. You could just extract the checksum from the file and then add it as a tag on the artifact like `input-sha256:abcdef0123456799...`. Then, your query can include that tag as a criterion (as well as any additional tags we may see fit to add in the future). I'm not sure if there's a length limit on tags as there appears to be on versions, but at that point, you can truncate the checksum to whatever length you need to, but you're not truncating it for any other future uses (because the full-length checksum is still available in the file).

Reply:
Unfortunately you cannot have two raw packages of the same version but different tags. The tag is a nonstarter here.

https://grapl-internal.slack.com/archives/C0174A8QV2S/p1647882336138319